src/marlin_stubs/pause/pause.cpp

```diff
@@ -709,6 +711,101 @@ bool Pause::wait_or_stop() {
     return false;
 }
 
+// TODO: Implement non-blocking homing, put it in homing.cpp and replace this
+static void prepare_for_homing() {
+
+    // Wait for planner moves to finish!
+    planner.synchronize();
+
+// Disable the leveling matrix before homing
+#if HAS_LEVELING
+
+    // Cancel the active G29 session
+    #if ENABLED(PROBE_MANUALLY)
+    g29_in_progress = false;
+    #endif
+
+    #if ENABLED(RESTORE_LEVELING_AFTER_G28)
+    const bool leveling_was_active = planner.leveling_active;
+    #endif
+    set_bed_leveling_enabled(false);
+#endif
+
+#if ENABLED(CNC_WORKSPACE_PLANES)
+    workspace_plane = PLANE_XY;
+#endif
+
+#if ENABLED(IMPROVE_HOMING_RELIABILITY)
+    slow_homing_t slow_homing { 0 };
+    slow_homing.acceleration.set(planner.settings.max_acceleration_mm_per_s2[X_AXIS],
+        planner.settings.max_acceleration_mm_per_s2[Y_AXIS]);
+    planner.settings.max_acceleration_mm_per_s2[X_AXIS] = 100;
+    planner.settings.max_acceleration_mm_per_s2[Y_AXIS] = 100;
+    #if HAS_CLASSIC_JERK
+    slow_homing.jerk_xy = planner.max_jerk;
+    planner.max_jerk.set(0, 0);
+    #endif
+
+    planner.reset_acceleration_rates();
+#endif
+
+// Always home with tool 0 active
+#if HOTENDS > 1
+    #if DISABLED(DELTA) || ENABLED(DELTA_HOME_TO_SAFE_ZONE)
+    const uint8_t old_tool_index = active_extruder;
+    #endif
+    tool_change(0, true);
+#endif
+
+#if HAS_DUPLICATION_MODE
+    extruder_duplication_enabled = false;
+#endif
+
+    remember_feedrate_scaling_off();
+    endstops.enable(true); // Enable endstops for next homing move
+    destination = current_position;
+}
+
+static void finalize_homing() {
+    endstops.not_homing();
+
+// Clear endstop state for polled stallGuard endstops
+#if ENABLED(SPI_ENDSTOPS)
+    endstops.clear_endstop_state();
+#endif
+
+#if HAS_LEVELING && ENABLED(RESTORE_LEVELING_AFTER_G28)
+    set_bed_leveling_enabled(leveling_was_active);
+#endif
+
+    restore_feedrate_and_scaling();
+
+// Restore the active tool after homing
+#if HOTENDS > 1 && (DISABLED(DELTA) || ENABLED(DELTA_HOME_TO_SAFE_ZONE))
+    #if EITHER(PARKING_EXTRUDER, DUAL_X_CARRIAGE)
+        #define NO_FETCH false // fetch the previous toolhead
+    #else
+        #define NO_FETCH true
+    #endif
+    tool_change(old_tool_index, NO_FETCH);
+#endif
+
+#if ENABLED(IMPROVE_HOMING_RELIABILITY)
+    planner.settings.max_acceleration_mm_per_s2[X_AXIS] = slow_homing.acceleration.x;
+    planner.settings.max_acceleration_mm_per_s2[Y_AXIS] = slow_homing.acceleration.y;
+    #if HAS_CLASSIC_JERK
+    planner.max_jerk = slow_homing.jerk_xy;
+    #endif
+    planner.reset_acceleration_rates();
+#endif
+}
+
+static void home_axis_in_pause(AxisEnum axis) {
+    prepare_for_homing();
```

Comment:
I would remove all of this code and rework G28 in its file and create the G28_no_parser.